src/main/kotlin/cacophony/automata/DFAJoin.kt

```diff
@@ -0,0 +1,83 @@
+package cacophony.automata
+
+import cacophony.automata.minimalization.via
+
+// Joins given list of the automata into a single DFA with accepting states marked
+// with a Result corresponding to the source DFAs.
+// Throws if any state would be marked by many results, i.e. when any two DFAs accept the same word.
+fun <State, Atom, Result> joinAutomata(
+    automataAndResults: List<Pair<GenericDFA<State, Atom, *>, Result>>,
+): GenericDFA<Int, Atom, Result?> {
+    if (automataAndResults.isEmpty()) throw IllegalArgumentException("Provided empty list of automaton")
+    val automata = automataAndResults.unzip().first
+
+    // Get result for the given state or null if it is not accepting.
+    // Throws if the result is ambiguous.
+    val resultForState = fun(states: List<State?>): Result? {
+        val accept = (automataAndResults zip states).filter { (pair, state) -> state != null && pair.first.isAccepting(state) }
+        if (accept.size > 1) throw IllegalArgumentException("Provided automata are ambiguous")
+        val res = accept.firstOrNull() ?: return null
+        return res.first.second
```

Comment:
```kotlin
return accept.firstOrNull()?.first?.second
```

Reply:
Coverage cries that second null check never goes to false : ' )

Reply:
alright then it can stay like this :D